### PR TITLE
MySQLI emulator

### PR DESCRIPTION
**Pull request by rnehra01**
- [x] ~~Setup new database if NOT present~~
- [x] ~~Setup a new database for attacker~~
- [x] ~~Update config to take inputs like `db_password`, `choose db`~~
- [x] ~~Create abstraction for `db_helper` and `sqli`~~
- [x] ~~Identify MySQLI, handle SQLI(for this no extra function need to be written, we just need to abstract in the previous task)~~

### CODE REVIEW

**Comment by rnehra01** [comment]
@afeena Any suggestions on what and how I've done.

File: tanner/emulators/mysqli.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import asyncio
+import pymysql
+import os
+import urllib.parse
+import pylibinjection
+from asyncio.subprocess import PIPE
```
Comment (afeena) [inline tanner/emulators/mysqli.py]: unused

**Comment by afeena** [comment]
@rnehra01 For first attempt it is great :) 
I highly recommend you to move toward to high level of abstraction. I mean we shouldn't copy the same methods for every db engine. 
Also we can give the possibility to choose the db in config, for example.

**Comment by rnehra01** [comment]
I have added abstraction in the list at `5` and `6` and will try to implement that soon.

File: tanner/emulators/mysqli.py (hunk 1 above)
Comment (rnehra01) [inline tanner/emulators/mysqli.py]: Actually imported all the things that will be used, came to my mind.

**Comment by rnehra01** [comment]
@afeena take a look at `copy_db` function.

File: tanner/utils/mysql_db_helper.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        
+    @asyncio.coroutine
+    def setup_db_from_config(self, name=None):
+        config = yield from self.read_config()
+        if name is not None:
+            db_name = name
+        else:
+            db_name = config['name']
+               
+        conn = connect_to_db()
```
Comment (afeena) [inline tanner/utils/mysql_db_helper.py]: `yield from self.connect_to_db()`

File: tanner/utils/mysql_db_helper.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+        conn.close()
+
+    @asyncio.coroutine
+    def copy_db(user_db, attacker_db):
+        if check_db_exists(attacker_db):
+            self.logger.info('Attacker db already exists')
+        else:
+            #create new attacker db
+            conn = connect_to_db()
```
Comment (afeena) [inline tanner/utils/mysql_db_helper.py]: `yield from self.connect_to_db()`

File: tanner/utils/mysql_db_helper.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            copy_db_cmd+= ' | '
+            copy_db_cmd+= restore_db_cmd.format(host='localhost', user='root', password='*********', db_name=attacker_db)
+            os.system(copy_db_cmd)
+
+    @asyncio.coroutine
+    def create_query_map(self,db_name, ):
+        query_map = {}
+        tables = []
+
+        conn = connect_to_db()
```
Comment (afeena) [inline tanner/utils/mysql_db_helper.py]: `yield from self.connect_to_db()`

**Comment by rnehra01** [comment]
Will resolve them when free.

**Comment by rnehra01** [comment]
@afeena resolved some of them.
Yeah, it will better to use `async library for mysql` because of its natural asynchronous nature and we have used asyncio a lot in the whole project.

**Comment by glaslos** [comment]
what is the status of this PR?

**Comment by rnehra01** [comment]
@glaslos I'm currently busy with my semester exams. I'll implement the remaining things after 29 Apr when my exams will end.

**Comment by rnehra01** [comment]
@afeena Almost finished implementing it. Please take a look. What I was thinking is to merge `mysqli.py` file with `mysly_db_helper` as former doesn't have much functions and all can be accommodated in the latter. And same for sqlite counter part.

**Comment by rnehra01** [comment]
Also can I replace sqlite db name with `tanner_db` instead of `tanner.db` so that it can match with mysql one.
Also I've prepended `attacker_` in mysql attacker db name as there is some problem in naming a database starting with integer(session id), so I'm thinking of doing same for sqlite.

**Comment by rnehra01** [comment]
Fixed delete db in MySQL, port to `await-async`syntax

**Comment by rnehra01** [comment]
@afeena will you merge them as a single commit or with all these, if latter let me know I'll squash some of commits named `fix` to make history look cleaner.

**Comment by glaslos** [comment]
@rnehra01 no worries, we can squash from the GitHub UI. It's usually good for us to see the individual commits so we can follow your changes.

**Comment by rnehra01** [comment]
Time based blind SQLI is supported by MySQL by default, #108

**Pull request merged**